src/send-message.ts

```diff
@@ -0,0 +1,176 @@
+}
+
+function parsePrivateMessageDestinations(
+  input: string
+): PrivateDestination["destination"] {
+  let rawDestinations = input.split(",");
+  return allNumeric(rawDestinations)
+    ? rawDestinations.map((item: string) => Number.parseInt(item))
+    : rawDestinations;
+}
+
+function parseStreamDestination(
+  input: string
+): StreamDestination["destination"] {
+  return allNumeric(input) ? Number.parseInt(input) : input;
+}
+
+function parseDestinationDetails({
+  kind,
+  destination,
+}: {
+  kind: DestinationKind;
+  destination: string;
+}): Result<DestinationDetails, string> {
+  switch (kind) {
+    case DestinationKind.Private: {
+      return Ok({
+        kind,
+        destination: parsePrivateMessageDestinations(destination),
+      });
+    }
+
+    case DestinationKind.Stream: {
+      return getMandatoryInputFromJob("topic")
+        .map((topic) => {
+          return {
+            kind,
+            topic,
+            destination: parseStreamDestination(destination),
+          };
+        })
+        .mapErr(() => 'topic is mandatory when type is "stream"');
+    }
+
+    default: {
+      assertExhaustiveSwitch(kind);
+    }
+  }
+}
+
+function getMandatoryInputFromJob(key: string): Result<string, string> {
+  const candidate = getInput(key, { required: true });
+
+  return candidate
+    ? Ok(candidate)
+    : Err(`input "${key}" must be provided and non-empty`);
+}
+
+// TypeScript lacks a way to genericize this cleanly to
+// getEnumInputFromGHA<T extends Enum>(key: string, spec: T),
+// see https://github.com/microsoft/TypeScript/issues/30611
+function getDestinationKindInputFromJob(): Result<DestinationKind, string> {
+  const validValues = Object.values(DestinationKind);
+  const errorMessage = `input "type" must exist and be one of ${validValues.join(
+    " ,"
+  )}`;
+  return getMandatoryInputFromJob("type")
+    .mapErr(() => errorMessage)
+    .andThen((input) => {
+      return validValues.includes(input as any)
+        ? Ok(input as DestinationKind)
+        : Err(errorMessage);
+    });
+}
+
+function getDestinationDetails(): Result<DestinationDetails, string> {
+  return getDestinationKindInputFromJob()
+    .andThen((destinationKind) => {
+      return getMandatoryInputFromJob("to").map((destination) => {
+        return { kind: destinationKind, destination };
+      });
+    })
+    .andThen(parseDestinationDetails);
+}
+
+async function getZulipClient(): Promise<Result<Client, string>> {
+  return new AsyncResultWrapper(
+    Result.all(
+      getMandatoryInputFromJob("api-key"),
+      getMandatoryInputFromJob("email"),
+      getMandatoryInputFromJob("organization-url")
+    )
+  )
+    .flatMap(
+      ([apiKey, username, realm]) =>
+        new AsyncResultWrapper(
+          Result.wrapAsync(async () =>
+            zulipInit({
+              apiKey,
+              username,
+              realm,
+            })
+          )
+        )
```

Comment:
Frankly this interaction between `ts-async-results` and the parent library, `ts-results`, looks pretty gross. There may be a simpler way to do this, but after reading upstream issues in both repos and a lengthy conversation with the type checker, this is the best I came up with so far.